src/realm/sync/noinst/client_impl_base.cpp

```diff
@@ -1887,14 +1888,20 @@ void Session::send_message()
             return false;
         }
 
-        return m_upload_progress.client_version >= m_pending_flx_sub_set->snapshot_version;
+        // Send QUERY messages when the upload progress client version reaches the snapshot version
+        // of a pending subscription, or if this is a fresh realm download session, since UPLOAD
+        // messages are not allowed and the upload progress will not be updated.
+        return m_upload_progress.client_version >= m_pending_flx_sub_set->snapshot_version ||
+               REALM_UNLIKELY(is_fresh_realm_download());
```

Comment:
I'm pretty sure this logic is invalid? We can't reorder things like this because the pending subscription may depend on the writes before it which are waiting to be uploaded. We don't want to send Pending subscriptions to the server  when downloading a fresh Realm; we want to send the currently active one.

Reply:
this is only for the fresh realm, so there are no meaningful writes. The fresh realm bootstraps using the active subscriptions of the local realm (but they are marked pending when committed to the fresh realm) so I think it's safe to assume that any changes before that were uploaded because otherwise we would have had the same problem when bootstrapping the local realm.

Reply:
The fresh realm sync session has its own (new) subscription store with only one pending subscription (other than query version 0), which is the active subscription from the local realm. Since we're not sending upload messages, we can't rely on the `m_upload_progress.client_version` value - I guess it could wait until query version 0 is complete.